mathics/builtin/structure.py

```diff
@@ -85,6 +88,138 @@ def __gt__(self, other):
             return Expression(list.head, *new_leaves)
 
 
+class SortBy(Builtin):
+    """
+    <dl>
+    <dt>'SortBy[$list$, $f$]'
+    <dd>sorts $list$ (or the leaves of any other expression) according to canonical ordering of the keys that are
+    extracted from the $list$'s elements using $f. Chunks of leaves that appear the same under $f are sorted
+    according to their natural order (without applying $f).
+    <dt>'SortBy[$f$]'
+    <dd>creates an operator function that, when applied, sorts by $f.
+    </dl>
+
+    >> SortBy[{{5, 1}, {10, -1}}, Last]
+    = {{10, -1}, {5, 1}}
+
+    >> SortBy[Total][{{5, 1}, {10, -9}}]
+    = {{10, -9}, {5, 1}}
+    """
+
+    rules = {
+        'SortBy[f_]': 'SortBy[#, f]&'
+    }
+
+    def apply(self, l, f, evaluation):
+        'SortBy[l_, f_]'
+
+        if l.is_atom():
+            evaluation.message('Sort', 'normal')
+        else:
+            keys = Expression('Map', f, l).evaluate(evaluation).leaves  # precompute:
+            # even though our sort function has only (n log n) comparisons, we should
+            # compute f no more than n times.
+
+            raw_keys = l.leaves
+            if len(raw_keys) != len(keys):  # this should never happen, I assume
+                return Symbol('$Aborted')
```

Comment:
I'd guess this can happen if `f` is not a valid function. We should probably check that `l` is actually a `List` and print an error message + return None if things go wrong.